onnxruntime/core/graph/graph_utils.cc

```diff
@@ -399,48 +399,80 @@ bool IsGraphInput(const Graph& graph, const NodeArg* input) {
   return std::find(graph_inputs.begin(), graph_inputs.end(), input) != graph_inputs.end();
 }
 
-bool IsConstantInitializer(const Graph& graph, const std::string& initializer_name, bool check_outer_scope) {
-  const onnx::TensorProto* initializer = nullptr;
-  bool is_local_initializer = graph.GetInitializedTensor(initializer_name, initializer);
-
-  // if we know it's an initializer we assume it's constant initially.
-  // otherwise it's not an initializer so can't be a constant initializer by definition.
-  bool constant_initializer = is_local_initializer;
-
-  if (is_local_initializer) {
+const ONNX_NAMESPACE::TensorProto* GetConstantInitializer(const Graph& graph, const std::string& initializer_name,
+                                                          bool check_outer_scope) {
+  const ONNX_NAMESPACE::TensorProto* initializer = nullptr;
+  if (graph.GetInitializedTensor(initializer_name, initializer)) {
     if (graph.CanOverrideInitializer()) {
       const auto& graph_inputs = graph.GetInputsIncludingInitializers();
-      constant_initializer = std::none_of(graph_inputs.cbegin(), graph_inputs.cend(),
-                                          [&initializer_name](const NodeArg* input) {
-                                            return input->Name() == initializer_name;
-                                          });
+      bool is_constant = std::none_of(graph_inputs.cbegin(), graph_inputs.cend(),
+                                      [&initializer_name](const NodeArg* input) {
+                                        return input->Name() == initializer_name;
+                                      });
+
+      if (!is_constant) {
+        initializer = nullptr;
+      }
+    } else if (check_outer_scope && graph.IsSubgraph()) {
+      initializer = GetConstantInitializer(*graph.ParentGraph(), initializer_name);
     }
-  } else if (check_outer_scope && graph.IsSubgraph()) {
-    constant_initializer = IsConstantInitializer(*graph.ParentGraph(), initializer_name, check_outer_scope);
   }
 
-  return constant_initializer;
+  return initializer;
+}
+
+bool IsConstantInitializer(const Graph& graph, const std::string& initializer_name, bool check_outer_scope) {
+  const onnx::TensorProto* initializer = GetConstantInitializer(graph, initializer_name, check_outer_scope);
+  return initializer != nullptr;
 }
 
 bool NodeArgIsConstant(const Graph& graph, const NodeArg& node_arg) {
   return IsConstantInitializer(graph, node_arg.Name(), true);
 }
 
-bool AllNodeInputsAreConstant(const Graph& graph, const Node& node) {
+bool AllNodeInputsAreConstant(const Graph& graph, const Node& node, InitializedTensorSet& constant_inputs) {
```

Comment:
It's slightly more flexible to allow the caller to add to an existing set if they wanted to so I'd rather leave the choice up to them.

Reply:
I hear you, but what then there is the `constant_inputs.clear()` after. So if the caller wants to give us a set with preexisting defs, we should either clear it always or not. Otherwise, we can get to a situation that we have the preexisting defs, plus some more that we added with this method, and then we see that not all are initializers. If we clear the list, the caller loses its preexisting defs too. So I would suggest to either clear the set always or store the preexisting and put them back in case not all are constants.

Reply:
good point. I'll clear at the start.